DebugConfig deserialization for agent/self

The call to `/agent/self` fails when using consul 1.8.4

```
    /**
     * Retrieves the Agent's configuration and member information.
     * <p/>
     * GET /v1/agent/self
     *
     * @return The Agent information.
     */
    public Agent getAgent() {
        return http.extract(api.getAgent());
    }
```

```
Caused by: java.lang.IllegalStateException: Cannot build DebugConfig, some of required attributes are not set [logLevel, enableSyslog]
        at com.orbitz.consul.model.agent.ImmutableDebugConfig$Builder.build(ImmutableDebugConfig.java:2121)
        at com.orbitz.consul.model.agent.ImmutableDebugConfig.fromJson(ImmutableDebugConfig.java:1586)
```

From the documentation, DebugConfig is an internal data structure and should is subject to change:

> This endpoint returns the configuration and member information of the local agent. The Config element contains a subset of the configuration and its format will not change in a backwards incompatible way between releases. DebugConfig contains the full runtime configuration but its format is subject to change without notice or deprecation. https://www.consul.io/api-docs/agent

All fields in `DebugConfig` are better made optional or nullable.

--- a/src/itest/java/com/orbitz/consul/AgentITest.java
+++ b/src/itest/java/com/orbitz/consul/AgentITest.java
@@ -36,18 +36,19 @@ public class AgentITest extends BaseIntegrationTest {
     private static final Map<String, String> NO_META = Collections.emptyMap();
 
     @Test
-    public void shouldRetrieveAgentInformation() throws UnknownHostException {
+    public void shouldRetrieveAgentInformation() {
         Agent agent = client.agentClient().getAgent();
 
         org.junit.Assume.assumeTrue(agent.getDebugConfig() != null);
 
         assertNotNull(agent);
         assertNotNull(agent.getConfig());
-        assertNotNull(agent.getDebugConfig().getClientAddrs().get(0));
+        final List<?> clientAddrs = (List<?>) agent.getDebugConfig().get("ClientAddrs");
+        assertNotNull(clientAddrs.get(0));
 
         // maybe we should not make any assertion on the actual value of the client address
         // as like when we run consul in a docker container we would have "0.0.0.0"
-        assertThat(agent.getDebugConfig().getClientAddrs().get(0), anyOf(is("127.0.0.1"), is("0.0.0.0")));
+        assertThat(clientAddrs.get(0), anyOf(is("127.0.0.1"), is("0.0.0.0")));
     }
 
     @Test

--- a/src/main/java/com/orbitz/consul/cache/ConsulCache.java
+++ b/src/main/java/com/orbitz/consul/cache/ConsulCache.java
@@ -130,7 +130,7 @@ public class ConsulCache<K, V> implements AutoCloseable {
                     ImmutableMap<K, V> full = convertToMap(consulResponse);
 
                     boolean changed = !full.equals(lastResponse.get());
-                    eventHandler.cachePollingSuccess(cacheDescriptor, changed, elapsedTime);
+                    eventHandler.cachePollingSuccess(cacheDescriptor, changed, Duration.ofMillis(elapsedTime));
 
                     if (changed) {
                         // changes
@@ -171,7 +171,7 @@ public class ConsulCache<K, V> implements AutoCloseable {
                             cacheConfig.getMinimumDurationDelayOnEmptyResult().compareTo(timeToWait) > 0) {
                         timeToWait = cacheConfig.getMinimumDurationDelayOnEmptyResult();
                     }
-                    timeToWait = timeToWait.minus(elapsedTime);
+                    timeToWait = timeToWait.minusMillis(elapsedTime);
 
                     scheduler.schedule(ConsulCache.this::runCallback,
                             timeToWait.toMillis(), TimeUnit.MILLISECONDS);

--- a/src/main/java/com/orbitz/consul/model/agent/Agent.java
+++ b/src/main/java/com/orbitz/consul/model/agent/Agent.java
@@ -7,6 +7,7 @@ import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import org.immutables.value.Value;
 
+import java.util.Map;
 import javax.annotation.Nullable;
 
 @Value.Immutable

--- a/src/main/java/com/orbitz/consul/model/agent/DebugConfig.java
+++ b/src/main/java/com/orbitz/consul/model/agent/DebugConfig.java
@@ -1,88 +1,10 @@
 package com.orbitz.consul.model.agent;
 
-import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
-import com.fasterxml.jackson.annotation.JsonInclude;
-import com.fasterxml.jackson.annotation.JsonProperty;
-import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
-import com.fasterxml.jackson.databind.annotation.JsonSerialize;
-import java.util.Optional;
-import org.immutables.value.Value;
+import java.util.HashMap;
 
-import java.util.List;
+/**
+ * DebugConfig contains the full runtime configuration but its format is subject to change without notice or deprecation.
+ */
+public class DebugConfig extends HashMap<String, Object> {
 
-@Value.Immutable
-@JsonSerialize(as = ImmutableDebugConfig.class)
-@JsonDeserialize(as = ImmutableDebugConfig.class)
-@JsonIgnoreProperties(ignoreUnknown = true)
-@JsonInclude(JsonInclude.Include.ALWAYS)
-public abstract class DebugConfig {
-
-    @JsonProperty("Bootstrap")
-    public abstract boolean getBootstrap();
-
-    @JsonProperty("Datacenter")
-    public abstract String getDatacenter();
-
-    @JsonProperty("DataDir")
-    public abstract String getDataDir();
-
-    @JsonProperty("DNSRecursors")
-    public abstract List<String> dnsRecursors();
-
-    @JsonProperty("DNSDomain")
-    public abstract String getDnsDomain();
-
-    @JsonProperty("LogLevel")
-    public abstract String getLogLevel();
-
-    @JsonProperty("NodeName")
-    public abstract String getNodeName();
-
-    @JsonProperty("ClientAddrs")
-    public abstract List<String> getClientAddrs();
-
-    @JsonProperty("BindAddr")
-    public abstract String getBindAddr();
-
-    @JsonProperty("LeaveOnTerm")
-    public abstract boolean getLeaveOnTerm();
-
-    @JsonProperty("SkipLeaveOnInt")
-    public abstract boolean getSkipLeaveOnInt();
-
-    @JsonProperty("EnableDebug")
-    public abstract boolean getEnableDebug();
-
-    @JsonProperty("VerifyIncoming")
-    public abstract boolean getVerifyIncoming();
-
-    @JsonProperty("VerifyOutgoing")
-    public abstract boolean getVerifyOutgoing();
-
-    @JsonProperty("CAFile")
-    public abstract String getCaFile();
-
-    @JsonProperty("CertFile")
-    public abstract String getCertFile();
-
-    @JsonProperty("KeyFile")
-    public abstract String getKeyFile();
-
-    @JsonProperty("UiDir")
-    public abstract Optional<String> getUiDir();
-
-    @JsonProperty("PidFile")
-    public abstract String getPidFile();
-
-    @JsonProperty("EnableSyslog")
-    public abstract boolean getEnableSyslog();
-
-    @JsonProperty("RejoinAfterLeave")
-    public abstract boolean getRejoinAfterLeave();
-
-    @JsonProperty("AdvertiseAddrLAN")
-    public abstract String getAdvertiseAddrLAN();
-
-    @JsonProperty("AdvertiseAddrWAN")
-    public abstract String getAdvertiseAddrWAN();
 }

--- a/src/test/java/com/orbitz/consul/model/agent/DebugConfigTest.java
+++ b/src/test/java/com/orbitz/consul/model/agent/DebugConfigTest.java
@@ -1,0 +1,245 @@
+package com.orbitz.consul.model.agent;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.orbitz.consul.util.Jackson;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+
+public class DebugConfigTest {
+    private final String JSON = "{\n" +
+            "    \"ACLMasterToken\": \"hidden\",\n" +
+            "    \"GRPCAddrs\": [\n" +
+            "        \"tcp://0.0.0.0:8502\"\n" +
+            "    ],\n" +
+            "    \"ACLEnableKeyListPolicy\": false,\n" +
+            "    \"HTTPSPort\": -1,\n" +
+            "    \"AutoEncryptTLS\": false,\n" +
+            "    \"AdvertiseAddrLAN\": \"127.0.0.1\",\n" +
+            "    \"HTTPAddrs\": [\n" +
+            "        \"tcp://0.0.0.0:8500\"\n" +
+            "    ],\n" +
+            "    \"LeaveOnTerm\": false,\n" +
+            "    \"ClientAddrs\": [\n" +
+            "        \"0.0.0.0\"\n" +
+            "    ],\n" +
+            "    \"RetryJoinIntervalLAN\": \"30s\",\n" +
+            "    \"ACLTokenTTL\": \"30s\",\n" +
+            "    \"ACLDefaultPolicy\": \"allow\",\n" +
+            "    \"CheckReapInterval\": \"30s\",\n" +
+            "    \"ConnectEnabled\": true,\n" +
+            "    \"ConsulCoordinateUpdatePeriod\": \"100ms\",\n" +
+            "    \"ConsulRaftHeartbeatTimeout\": \"35ms\",\n" +
+            "    \"AutoEncryptAllowTLS\": false,\n" +
+            "    \"RaftSnapshotThreshold\": 0,\n" +
+            "    \"ACLTokenReplication\": false,\n" +
+            "    \"GRPCPort\": 8502,\n" +
+            "    \"AutopilotDisableUpgradeMigration\": false,\n" +
+            "    \"BindAddr\": \"127.0.0.1\",\n" +
+            "    \"VerifyIncomingRPC\": false,\n" +
+            "    \"DNSAddrs\": [\n" +
+            "        \"tcp://0.0.0.0:8600\",\n" +
+            "        \"udp://0.0.0.0:8600\"\n" +
+            "    ],\n" +
+            "    \"ACLReplicationToken\": \"hidden\",\n" +
+            "    \"AutopilotMinQuorum\": 0,\n" +
+            "    \"SerfAdvertiseAddrLAN\": \"tcp://127.0.0.1:8301\",\n" +
+            "    \"ConsulRaftLeaderLeaseTimeout\": \"20ms\",\n" +
+            "    \"UIContentPath\": \"/ui/\",\n" +
+            "    \"EnableCentralServiceConfig\": false,\n" +
+            "    \"KVMaxValueSize\": 524288,\n" +
+            "    \"ServerMode\": true,\n" +
+            "    \"TxnMaxReqLen\": 524288,\n" +
+            "    \"DisableRemoteExec\": true,\n" +
+            "    \"HTTPMaxConnsPerClient\": 200,\n" +
+            "    \"RPCHandshakeTimeout\": \"5s\",\n" +
+            "    \"DNSMaxStale\": \"87600h0m0s\",\n" +
+            "    \"Telemetry\": {\n" +
+            "        \"CirconusAPIToken\": \"hidden\",\n" +
+            "        \"Disable\": false,\n" +
+            "        \"DisableHostname\": false,\n" +
+            "        \"FilterDefault\": true,\n" +
+            "        \"MetricsPrefix\": \"consul\",\n" +
+            "        \"PrometheusRetentionTime\": \"0s\"\n" +
+            "    },\n" +
+            "    \"SegmentNameLimit\": 64,\n" +
+            "    \"GossipLANGossipNodes\": 3,\n" +
+            "    \"RaftTrailingLogs\": 0,\n" +
+            "    \"RPCProtocol\": 2,\n" +
+            "    \"DNSCacheMaxAge\": \"0s\",\n" +
+            "    \"GossipLANRetransmitMult\": 4,\n" +
+            "    \"AutopilotMaxTrailingLogs\": 250,\n" +
+            "    \"ACLDatacenter\": \"dc1\",\n" +
+            "    \"DefaultQueryTime\": \"5m0s\",\n" +
+            "    \"Version\": \"1.8.4\",\n" +
+            "    \"RPCRateLimit\": -1,\n" +
+            "    \"Logging\": {\n" +
+            "        \"EnableSyslog\": false,\n" +
+            "        \"LogJSON\": false,\n" +
+            "        \"LogLevel\": \"DEBUG\",\n" +
+            "        \"LogRotateBytes\": 0,\n" +
+            "        \"LogRotateDuration\": \"0s\",\n" +
+            "        \"LogRotateMaxFiles\": 0,\n" +
+            "        \"SyslogFacility\": \"LOCAL0\"\n" +
+            "    },\n" +
+            "    \"RPCMaxConnsPerClient\": 100,\n" +
+            "    \"RejoinAfterLeave\": false,\n" +
+            "    \"GossipWANSuspicionMult\": 3,\n" +
+            "    \"DNSNodeMetaTXT\": true,\n" +
+            "    \"ACLsEnabled\": false,\n" +
+            "    \"SerfPortWAN\": 8302,\n" +
+            "    \"GossipLANProbeInterval\": \"100ms\",\n" +
+            "    \"PrimaryDatacenter\": \"dc1\",\n" +
+            "    \"Cache\": {\n" +
+            "        \"EntryFetchMaxBurst\": 2,\n" +
+            "        \"EntryFetchRate\": 1.7976931348623157E308\n" +
+            "    },\n" +
+            "    \"DiscardCheckOutput\": false,\n" +
+            "    \"ReconnectTimeoutWAN\": \"0s\",\n" +
+            "    \"AdvertiseAddrWAN\": \"127.0.0.1\",\n" +
+            "    \"ConnectSidecarMaxPort\": 21255,\n" +
+            "    \"EnableUI\": true,\n" +
+            "    \"EnableDebug\": true,\n" +
+            "    \"KeyFile\": \"hidden\",\n" +
+            "    \"RPCAdvertiseAddr\": \"tcp://127.0.0.1:8300\",\n" +
+            "    \"SyncCoordinateIntervalMin\": \"15s\",\n" +
+            "    \"EncryptVerifyOutgoing\": true,\n" +
+            "    \"DNSUseCache\": false,\n" +
+            "    \"Bootstrap\": false,\n" +
+            "    \"TranslateWANAddrs\": false,\n" +
+            "    \"ACLEnableTokenPersistence\": false,\n" +
+            "    \"DNSNodeTTL\": \"0s\",\n" +
+            "    \"ExposeMinPort\": 21500,\n" +
+            "    \"DNSDomain\": \"consul.\",\n" +
+            "    \"SegmentLimit\": 64,\n" +
+            "    \"AutopilotLastContactThreshold\": \"200ms\",\n" +
+            "    \"ConsulCoordinateUpdateMaxBatches\": 5,\n" +
+            "    \"EnableRemoteScriptChecks\": true,\n" +
+            "    \"DNSAllowStale\": true,\n" +
+            "    \"AutopilotCleanupDeadServers\": true,\n" +
+            "    \"GossipWANGossipNodes\": 3,\n" +
+            "    \"NodeName\": \"484c56552d86\",\n" +
+            "    \"HTTPUseCache\": true,\n" +
+            "    \"DNSDisableCompression\": false,\n" +
+            "    \"SyncCoordinateRateTarget\": 64,\n" +
+            "    \"SerfAdvertiseAddrWAN\": \"tcp://127.0.0.1:8302\",\n" +
+            "    \"VerifyServerHostname\": false,\n" +
+            "    \"GossipWANProbeInterval\": \"100ms\",\n" +
+            "    \"ExposeMaxPort\": 21755,\n" +
+            "    \"VerifyOutgoing\": false,\n" +
+            "    \"DevMode\": true,\n" +
+            "    \"GossipWANProbeTimeout\": \"100ms\",\n" +
+            "    \"RetryJoinMaxAttemptsWAN\": 0,\n" +
+            "    \"DNSARecordLimit\": 0,\n" +
+            "    \"EncryptVerifyIncoming\": true,\n" +
+            "    \"DisableHTTPUnprintableCharFilter\": false,\n" +
+            "    \"GossipLANSuspicionMult\": 3,\n" +
+            "    \"DisableAnonymousSignature\": true,\n" +
+            "    \"SerfBindAddrWAN\": \"tcp://127.0.0.1:8302\",\n" +
+            "    \"ACLDownPolicy\": \"extend-cache\",\n" +
+            "    \"ConsulServerHealthInterval\": \"10ms\",\n" +
+            "    \"LeaveDrainTime\": \"5s\",\n" +
+            "    \"PrimaryGatewaysInterval\": \"30s\",\n" +
+            "    \"GossipWANRetransmitMult\": 4,\n" +
+            "    \"DisableCoordinates\": false,\n" +
+            "    \"EnableAgentTLSForChecks\": false,\n" +
+            "    \"EnableLocalScriptChecks\": true,\n" +
+            "    \"DisableHostNodeID\": true,\n" +
+            "    \"RPCBindAddr\": \"tcp://127.0.0.1:8300\",\n" +
+            "    \"ACLToken\": \"hidden\",\n" +
+            "    \"DataDir\": \"/consul/data\",\n" +
+            "    \"BootstrapExpect\": 0,\n" +
+            "    \"ConsulCoordinateUpdateBatchSize\": 128,\n" +
+            "    \"HTTPSHandshakeTimeout\": \"5s\",\n" +
+            "    \"EncryptKey\": \"hidden\",\n" +
+            "    \"ACLDisabledTTL\": \"2m0s\",\n" +
+            "    \"SerfPortLAN\": 8301,\n" +
+            "    \"ACLRoleTTL\": \"0s\",\n" +
+            "    \"ServerPort\": 8300,\n" +
+            "    \"ConnectTestCALeafRootChangeSpread\": \"0s\",\n" +
+            "    \"VerifyIncomingHTTPS\": false,\n" +
+            "    \"DiscoveryMaxStale\": \"0s\",\n" +
+            "    \"GossipLANGossipInterval\": \"100ms\",\n" +
+            "    \"GossipLANProbeTimeout\": \"100ms\",\n" +
+            "    \"TLSPreferServerCipherSuites\": false,\n" +
+            "    \"MaxQueryTime\": \"10m0s\",\n" +
+            "    \"ReconnectTimeoutLAN\": \"0s\",\n" +
+            "    \"CheckUpdateInterval\": \"5m0s\",\n" +
+            "    \"TLSMinVersion\": \"tls12\",\n" +
+            "    \"AutoConfig\": {\n" +
+            "        \"Authorizer\": {\n" +
+            "            \"AllowReuse\": false,\n" +
+            "            \"AuthMethod\": {\n" +
+            "                \"Config\": {\n" +
+            "                    \"ClockSkewLeeway\": 0,\n" +
+            "                    \"ExpirationLeeway\": 0,\n" +
+            "                    \"NotBeforeLeeway\": 0\n" +
+            "                },\n" +
+            "                \"MaxTokenTTL\": \"0s\",\n" +
+            "                \"Name\": \"Auto Config Authorizer\",\n" +
+            "                \"RaftIndex\": {\n" +
+            "                    \"CreateIndex\": 0,\n" +
+            "                    \"ModifyIndex\": 0\n" +
+            "                },\n" +
+            "                \"Type\": \"jwt\"\n" +
+            "            },\n" +
+            "            \"Enabled\": false\n" +
+            "        },\n" +
+            "        \"Enabled\": false,\n" +
+            "        \"IntroToken\": \"hidden\"\n" +
+            "    },\n" +
+            "    \"ACLAgentToken\": \"hidden\",\n" +
+            "    \"DNSEnableTruncate\": false,\n" +
+            "    \"DisableKeyringFile\": true,\n" +
+            "    \"SkipLeaveOnInt\": true,\n" +
+            "    \"RaftProtocol\": 0,\n" +
+            "    \"CheckDeregisterIntervalMin\": \"1m0s\",\n" +
+            "    \"DNSRecursorTimeout\": \"2s\",\n" +
+            "    \"RPCMaxBurst\": 1000,\n" +
+            "    \"Revision\": \"12b16df32\",\n" +
+            "    \"CheckOutputMaxSize\": 4096,\n" +
+            "    \"TaggedAddresses\": {\n" +
+            "        \"lan\": \"127.0.0.1\",\n" +
+            "        \"lan_ipv4\": \"127.0.0.1\",\n" +
+            "        \"wan\": \"127.0.0.1\",\n" +
+            "        \"wan_ipv4\": \"127.0.0.1\"\n" +
+            "    },\n" +
+            "    \"AutopilotServerStabilizationTime\": \"10s\",\n" +
+            "    \"DNSOnlyPassing\": false,\n" +
+            "    \"VerifyIncoming\": false,\n" +
+            "    \"DNSSOA\": {\n" +
+            "        \"Expire\": 86400,\n" +
+            "        \"Minttl\": 0,\n" +
+            "        \"Refresh\": 3600,\n" +
+            "        \"Retry\": 600\n" +
+            "    },\n" +
+            "    \"GossipWANGossipInterval\": \"100ms\",\n" +
+            "    \"ACLPolicyTTL\": \"30s\",\n" +
+            "    \"DisableUpdateCheck\": false,\n" +
+            "    \"DNSPort\": 8600,\n" +
+            "    \"NodeID\": \"46c53c05-5d11-8e08-c49b-a1e0c4a485b2\",\n" +
+            "    \"RetryJoinIntervalWAN\": \"30s\",\n" +
+            "    \"SessionTTLMin\": \"0s\",\n" +
+            "    \"NonVotingServer\": false,\n" +
+            "    \"ConnectMeshGatewayWANFederationEnabled\": false,\n" +
+            "    \"SerfBindAddrLAN\": \"tcp://127.0.0.1:8301\",\n" +
+            "    \"RetryJoinMaxAttemptsLAN\": 0,\n" +
+            "    \"ACLAgentMasterToken\": \"hidden\",\n" +
+            "    \"HTTPPort\": 8500,\n" +
+            "    \"ConsulRaftElectionTimeout\": \"52ms\",\n" +
+            "    \"RaftSnapshotInterval\": \"0s\",\n" +
+            "    \"ConnectSidecarMinPort\": 21000,\n" +
+            "    \"DNSUDPAnswerLimit\": 3,\n" +
+            "    \"AEInterval\": \"1m0s\",\n" +
+            "    \"Datacenter\": \"dc1\",\n" +
+            "    \"RPCHoldTimeout\": \"7s\"\n" +
+            "}";
+    @Test
+    public void testDeserialization() throws IOException {
+       ObjectMapper mapper = Jackson.MAPPER;
+       final DebugConfig dbgConfig = mapper.readerFor(DebugConfig.class).readValue(JSON);
+       assertEquals("DebugConfig contains 167 items", 167, dbgConfig.size());
+    }
+}

